### PR TITLE
Fix SiteNotice re-renders due to changing messagePath

### PR DESCRIPTION
**Pull request by getdave**
The purpose of this PR is exactly the same as https://github.com/Automattic/wp-calypso/pull/46676/. Please read that PR desc. Addresses part of https://github.com/Automattic/wp-calypso/issues/46091.

_This_ PR is the 3rd attempt (or option) for solving the same problem and largely mirrors @mreishus's [alternative approach](https://github.com/Automattic/wp-calypso/pull/46688) which limited the complexity by wrapping the `<JITM>` component and hiding the process of creating the `messagePath` and requiring only a `messagePathSuffix` prop.

The key difference in the approach in this PR is that we utilise a Higher Order Component which wraps the `<JITM>` component to remove the need for an additional `JITMLoader` component. Using an additional loader component required the consumer to need to know too much about the implementation details of `JITM`. Moreover, as the loader component was not the default export of the `blocks/jitm` directory it would be all too easy to import the wrong component.

By using the HOC approach we:

* Avoid the need to know about a "loader" component - the default export "just works".
* The existing convention of `calypso:${sectionName}:${messagePathSuffix}` is provided by default and codified within the HOC.
* It is still possible to use a customised `messagePath` (ie: one not based on `secitonName`) by importing the non-default export of `<JITMConnected>`.

The new component is consumed as follows:

```jsx

<AsyncLoad
    require="calypso/blocks/jitm"
    messagePathSuffix="sidebar_notice"
    template="sidebar-banner"
/>
```

#### Changes proposed in this Pull Request

* Alter props API of the default `<JITM>` component to require `messagePathSuffix`.
* Create HOC to abstract away building of `messagePath` to within the HOC - codify the existing convention.

#### Testing instructions

As per https://github.com/Automattic/wp-calypso/pull/46676/

### CODE REVIEW

**Review by cpapazoglou — APPROVED** [review]
This changes work well for me and remove the implementation details of the `JITM` from the `notice` component.

**Review by mreishus — APPROVED** [review]
This works in my testing, and improves upon #46676 and #46688, well done!

**Comment by sixhours** [comment]
This also works in testing for me! 🎉

**Comment by getdave** [comment]
Note that @jsnajdr suggested this change might not be required to fix the "jumpiness" of the sidebar menu in Calypso. He is actually right as https://github.com/Automattic/wp-calypso/pull/46677 will fix the jumpiness without the changes in this PR.

However, given that this PR makes it easier to use `JITM` without having to understand how to build `messagePath` it still might be worth merging.

I'll leave this up to others to pursue next week.

**Comment by mreishus** [comment]
I'll merge this in on Monday (2020-10-26) as Dave won't be available.

**Review by wp-desktop — CHANGES_REQUESTED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@mreishus please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/934050). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Review by wp-desktop — CHANGES_REQUESTED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@mreishus please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/934084). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

File: client/layout/index.jsx [hunk 1]
```diff
@@ -184,11 +184,7 @@ class Layout extends Component {
 				{ this.props.isOffline && <OfflineStatus /> }
 				<div id="content" className="layout__content">
 					{ config.isEnabled( 'jitms' ) && this.props.isEligibleForJITM && (
-						<AsyncLoad
-							require="calypso/blocks/jitm"
-							placeholder={ null }
-							messagePath={ `calypso:${ this.props.sectionJitmPath }:admin_notices` }
-						/>
+						<AsyncLoad require="calypso/blocks/jitm" messagePathSuffix="admin_notices" />
```
Comment (jsnajdr) [inline client/layout/index.jsx:187]: This change introduces a bug. For routes like `/themes/upload` or `/plugins/upload`, the `messagePath` used to be `calypso:themes-upload:admin_notices` and `calypso:plugins-upload:admin_notices`, respectively. The `themes-upload` slug was coming from the `getMessagePathForJITM` selector.

After this change, the `messagePath` is based on section name, the slug becoming `themes` and `plugins`.

The `themes-upload` and `plugins-upload` slugs are really used by the server and specific messages are generated:

Screenshot 2020-10-27 at 08 21 01

This place should use the `JITMConnected` variant exported from the `block/jitm` module, and supply a custom `messagePath`.

The trouble is that `AsyncLoad` can use only default exports of a module 🙁 An elegant solution would be to change the API of the `blocks/jitm` component: if the `messagePath` prop is supplied, use it, and generate one from `messagePathSuffix` only if it's not present:
```js
connect( ( state, { messagePath, messagePathSuffix } ) => {
  if ( messagePath === undefined ) {
    messagePath = `calypso:${ getSectionName( state ) }:${ messagePathSuffix }`;
  }
  return { messagePath };
} )
```
This is a fairly common Redux pattern.

Comment (mreishus) [reply client/layout/index.jsx:187]: Created #46821

Comment (mreishus) [reply client/layout/index.jsx:187]: PR: #46824

**Pull request merged**